### DISCUSSION TITLE
How shall I define a declarative HttpClient for a REST service with an operation taking a multi-valued query parameter?

### DISCUSSION THREAD

**Discussion by nedelva** [micronaut-projects/micronaut-core#10420] [Q&A]
I have defined this declarative client:
```java
@Client("${micronaut.http.services.my-api.url}")
public interface ContractsApi {
    @Get("/auto")
    @Consumes
    @Produces
    public List<Contracts> listContracts(@Nullable @QueryValue("Enterprise_Nbr") String enterpriseNbr,
                               @QueryValue("producers") @Format("csv") @Nullable List<String> producers);
}
```
All fine and dandy until a want to use the multi-valued query parameter producers. 
My issue here is that when passing a list of producers the generated URI looks like
`
/auto?Enterprise_Nbr=1234&producers=111%2C222%2C333
`

What the server expects is instead
`
/auto?Enterprise_Nbr=1234&producers=111,222,333
`
According to the user's guide (section 7.5.1 Customizing Parameter Binding), I should be able to use the Format annotation together with QueryValue to specify how the query values are to be formatted.

(I also tried to pass producers as a single String value by joining the values with comma delimiter, but the comma is escaped to %2C prior to the call)

**Comment by nedelva** [comment]
Not any Micronaut gurus out there willing to answer? If it cannot be achieved with the declarative client at least give me an alternative..
Passing an list of values in a query parameter seem such a basic feature, I don't believe it is not supported.

**Reply by graemerocher** [reply]
This is documented here https://docs.micronaut.io/latest/guide/#clientParameters

**Reply by nedelva** [reply]
Yes, Graeme, thanks for pointing that out. The section 7.5.1 I cited comes from that same location. Here's the relevant bit (and I quote):

> The [Format](https://docs.micronaut.io/latest/api/io/micronaut/core/convert/format/Format.html) annotation can be used together with @QueryValue annotation to format query values.

> The supported values are: "csv", "ssv", "pipes", "multi" and "deep-object", where the meaning is similar to [Open API v3](https://swagger.io/docs/specification/serialization/) query parameter’s style attribute.

> The format can only be applied to java.lang.Iterable, java.util.Map or POJO with [Introspected](https://docs.micronaut.io/latest/api/io/micronaut/core/annotation/Introspected.html) annotation. Examples of how different values will be formatted are given in the table below:


Format | Iterable example | Map or POJO example
-- | -- | --
Original value | ["Mike", "Adam", "Kate"] | {"name": "Mike", "age": 30"}
"CSV" | "param=Mike,Adam,Kate" | "param=name,Mike,age,30"
"SSV" | "param=Mike Adam Kate" | "param=name Mike age 30"
"PIPES" | "param=Mike\|Adam\|Kate" | "param=name\|Mike\|age\|30"
"MULTI" | "param=Mike&param=Adam&param=Kate" | "name=Mike&age=30"
"DEEP_OBJECT" | "param[0]=Mike&param[1]=Adam&param[2]=Kate" | "param[name]=Mike&param[age]=30"


However, the generated URI contains encoded commas (as I already mentioned in my first post) and I don't want that. The annotation _@Format("csv")_ does not have any effect in this particular setup.

**Reply by graemerocher** [reply]
Perhaps https://github.com/micronaut-projects/micronaut-core/blob/b106b5a02d03b88a5b2aa8a115bdec056fc9ded4/http-client/src/test/groovy/io/micronaut/http/client/format/ClientFormatSpec.groovy#L20 can shed some more light

**Reply by nedelva** [reply]
After some digging in the http-client codebase I ended up writing my own JUnit 5 version of the ClientFormatSpec (which ALSO passed all the tests, yay!).

Thing is, after setting up the logging level to DEBUG for http-client:
`<logger name="io.micronaut.http.client" level="debug" />`

I could see this output from running the tests:
```
09:39:55.971 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?param=name%7CMark%7Cage%7C30
09:39:56.000 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?param=tree%7Cspruce%7Cseed
09:39:56.010 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?param%5B0%5D=a&param%5B1%5D=b&param%5B2%5D=c%2Cd
09:39:56.016 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?param%5Bname%5D=Bob&param%5Bage%5D=8&param%5Bheight%5D=120
09:39:56.023 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?param%5Bname%5D=AAA&param%5Baddress%5D=here
09:39:56.032 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?ssv=a+b+c+d+e%2Cf%2Cg
09:39:56.037 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?ssv=name+Bob+age+8+height+120
09:39:56.041 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?ssv=name+AAA+address+here
09:39:56.055 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?param=tree%2Cspruce%2Cseed
09:39:56.066 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?param=name%7CPizza+Garden%7Caddress%7CHome+Street+2
09:39:56.077 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?multi=a&multi=b&multi=c%2Cd
09:39:56.085 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?name=Bob&age=8&height=120
09:39:56.092 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?name=AAA&address=here
09:39:56.103 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?csv=a%2Cb%2Cc
09:39:56.109 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?csv=name%2CBob%2Cage%2C12
09:39:56.115 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?csv=name%2CAAA%2Caddress%2Chere
09:39:56.125 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?param=a%7Cb%7Cc%7Cd%7Ce%2Cf%2Cg
09:39:56.130 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?param=name%7CBob%7Cage%7C8%7Cheight%7C120
09:39:56.134 [Test worker] DEBUG i.m.h.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:55952/format/queryIdentity?param=name%7CAAA%7Caddress%7Chere
```
So I guess the problem lies with `DefaultHttpClient`?
